Robust current_page? compare

bradhuskins opened the pull request:
### Motivation / Background

This MR updates UrlHelper#current_page? to compare target {:controller, :action} that will be routed to as opposed to comparison of URL strings.
To this end, we have added a couple of new methods to RouteSet (which is already loaded as a part of the request) to parse the 2 resultant URLs and determine controller and action names to be targeted [descoping the check to verify they are actual classes].
The usage of these new methods is restricted to the root route since, as pointed out in review, broader application is problematic.

FIXES #47184

### Detail

This Pull Request adds lighter weight path recognition to route_set, and then uses this to compare target controller/action combinations of the 2 URLs in the current_page? check, rather than just raw string comparison.

### Additional information

To test this, the test case from [Issue 47184](https://github.com/rails/rails/issues/47184), along with a few others, were imported into url_helper_test.rb, to ensure consistency.

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

bradhuskins committed:
Added detailed route/action comparison to current_page? helper

bradhuskins committed:
Added forgotten minitest helper

bradhuskins committed:
- cleanup

skipkayhil reviewed (commented):
What would happen with this code when a controller/action combination is used in multiple places in the route tree?

```ruby
resources :comments

resources :posts do
  resources :comments
end
```

Wouldn't this new implementation return true for both `/posts/:id/comments` and `/comments`?

p8 commented on activerecord/test/current_page_test.rb:
```diff
@@ -0,0 +1,32 @@
+    @request = request_for_url("/posts", method: :head)
+    assert current_page?(controller: "posts", action: "index")
+  end
+end
```
I think you probably want to add a test to the existing `current_page?` tests in ActionView: https://github.com/rails/rails/blob/64c8340bcbfcb3c91c1bf60c4b3de84c9d577bbe/actionview/test/template/url_helper_test.rb#L821

bradhuskins replied:
See comment below

pedro108 commented on activerecord/test/current_page_test.rb:
```diff
@@ -0,0 +1,32 @@
+require "minitest/autorun"
+require "action_dispatch/journey"
+
+class BugTest < ActiveSupport::TestCase
```
I believe that it makes sense to add a test to ensure that the [case mentioned](https://github.com/rails/rails/pull/47341#pullrequestreview-1292200782) by @skipkayhil does not result in a false positive

bradhuskins replied:
Added to url_helper_test

bradhuskins committed:
PR feedback:
- let's restrict the scope of our route equivalency check to
root override only
- move our tests in with rest of url_helpers

bradhuskins committed:
Linter cleanup

bradhuskins commented:
> 

Thanks for the feedback! We're not supposed to create Drafts apparently, otherwise this would have been.
I have restricted the scope of the "route equivalency" check to only be applied when the root route is requested without parameters, since this was really the crux of the issue created.

bradhuskins commented:
@skipkayhil Any thoughts on this scaled back approach, where we only perform the route equivalency check if the path passed maps to '/'? I believe it addresses your concerns.
